Review3

Shafagh-Sepehr opened the pull request.

Shafagh-Sepehr committed:
add

Shafagh-Sepehr committed:
move hard coded path to a newly created appsettings.json file.

Shafagh-Sepehr committed:
create new exception for null or whitespace input.

Shafagh-Sepehr committed:
create UserOutput class for output related actions.

Shafagh-Sepehr committed:
create UserInput class for input related actions.

Shafagh-Sepehr committed:
refactor UserInput and UserOutput classes name and methods and also extract interface.

Shafagh-Sepehr committed:
replace appsettings.json with a public static class for better accessibility across the project.

Shafagh-Sepehr committed:
move hardcoded banned words to the AppSettings class.

Shafagh-Sepehr committed:
refactor StringToWordsProcessor's TrimSplitAndStemString function:
extract two methods from it
get the stemmer via constructor instead of this method

Shafagh-Sepehr committed:
extract methods from TrimSplitAndStemString method

Shafagh-Sepehr committed:
change folder structure

Shafagh-Sepehr committed:
move InvertedIndex folder to a new Application folder.

Shafagh-Sepehr committed:
extract words processing from QuerySearcher.cs

Shafagh-Sepehr committed:
change QueryProcessor's name to WordProcessor

Shafagh-Sepehr committed:
refactored Query searcher.
It now inputs the inverted index through constructor.

Shafagh-Sepehr committed:
extract class from QuerySearcher.cs
GetDocumentForXQuery methods are now each a class and there's a master DocumentReader class that encapsulates them.

Shafagh-Sepehr committed:
extract methods from QuerySearcher's Search method

Shafagh-Sepehr committed:
improve inject IDocumentReader to QuerySearcher class for future test mocks.

Shafagh-Sepehr committed:
rename WordProcessor folder to WordProcessors because a class with the same name exists in it.

Shafagh-Sepehr committed:
fix the bug in QuerySearcher where Words are passed to document readers before initialization.

Shafagh-Sepehr committed:
extract methods from Search method of QuerySearcher class.

Shafagh-Sepehr committed:
rename Search methods to better ones.

Shafagh-Sepehr committed:
extract method from InvertedIndexDictionaryBuilder's FillInvertedIndexFromFile method.

Shafagh-Sepehr committed:
create builder class for InvertedIndexDictionary to decouple it from dependencies and also making instances easier.

Shafagh-Sepehr committed:
move regex patterns to AppSettings class

Shafagh-Sepehr committed:
Add Dependency Injection to the project for all classes.
create a Construct method to pass the needed parameters.
remove unused InvertedIndexBuilder.cs.
info: tests are down.

Shafagh-Sepehr committed:
Capitalize first letter of the Message in NullInputException.

Shafagh-Sepehr committed:
Add Assertions for classes that their Construct method should be called.

Shafagh-Sepehr committed:
extract Search methods to classes in QuerySearcher

Shafagh-Sepehr committed:
delete unused using statements.

Shafagh-Sepehr committed:
encapsulate class parameters

Shafagh-Sepehr committed:
remove interface postfix from interface namespaces.

Shafagh-Sepehr committed:
create InvertedIndexFactory for ease of use by both the user and developer.
extract Services to its own class.

Shafagh-Sepehr committed:
hide classes from user by changing them from public to internal.

Shafagh-Sepehr committed:
refactor: rename all "Interfaces" named folders to "Services"

Shafagh-Sepehr committed:
feat(exceptions): implement public constructors of each exception for handling of exceptions.

Shafagh-Sepehr committed:
refactor(AndDocumentsReader): for intersecting, use foreach loop instead of for loop and extract that part as a method.

Shafagh-Sepehr committed:
refactor: turn all private properties into readonly fields

Shafagh-Sepehr committed:
fix: add null checkers for all injected services.

Shafagh-Sepehr committed:
perf(searchers): change set operations methods to use HashSet methods.

convert linq's Except and Intersect methods to IntersectWith and ExceptWith methods of HashSet<T>.

Shafagh-Sepehr committed:
perf(Words): change Words.cs field types from List to IReadOnlyList.

this affects words-processors which create them and document-readers which use them.

Shafagh-Sepehr committed:
perf(Searchers): change return value types from IEnumerable to IReadOnlySet.

Shafagh-Sepehr committed:
perf(Words): seal Words class to gain performance and preventing inheritance.

Shafagh-Sepehr committed:
perf(AndDocumentsReader): make IntersectAllElements class static.

Shafagh-Sepehr committed:
refactor(StringToWordProcessor): replace long linq expression to several methods for better readability and maintainability.

Shafagh-Sepehr committed:
feat(appsettings): create appsettings.json file for user to maintain custom configurations.

Shafagh-Sepehr committed:
refactor: make all classes sealed because none of them was meant to be inherited

navidnasyry commented on FullTextSearch/Application/WordsProcessors/Services/IAndWordsProcessor.cs:
اینا سرویس نیستن...پیاده‌سازی ها سرویسن
اینا میشن abstraction ها مون
یعنی اگه میخوای پوشه‌بندیه خوب بشه
اینترفیس‌ها توی یه پوشه به نام Abstraction و پیاده‌سازی‌هاشون توی Service قرار میگیره

navidnasyry commented on FullTextSearch/ConfigHolder.cs:
write Service

navidnasyry replied:
and inject

navidnasyry commented on FullTextSearch/Services.cs:
```diff
@@ -0,0 +1,41 @@
+
+namespace FullTextSearch;
+
+internal static class Services
```
better class name
DICollection
DIServices
ServiceCollection
...

navidnasyry commented on FullTextSearch/Application/WordsProcessors/PrefixBasedAndWordsProcessor.cs:
```diff
@@ -0,0 +1,16 @@
+
+namespace FullTextSearch.Application.WordsProcessors;
+
+internal sealed class PrefixBasedAndWordsProcessor(IPorter2Stemmer stemmer) : IAndWordsProcessor
```
چرا ورودی کانستراکتور نمیگیری؟

navidnasyry replied:
همه کلاس‌هایی که میخوان از DI ابجکت بگیرن

navidnasyry commented on FullTextSearch/Application/Searchers/DataViewModels/Words.cs:
بهتره اسم پوشه Models باشه و اسم کلاسش دقیق‌تر باشه

navidnasyry commented on FullTextSearch/Application/InvertedIndex/StringToWordsProcessor.cs:
```diff
@@ -0,0 +1,73 @@
+
+    public IEnumerable<string> TrimSplitAndStemString(string source)
+    {
+        var result = CleanAndSplit(source);
+        result = PurgeNoise(result);
+        result = CleanAndSelect(result);
+        result = Stem(result);
+        result = PurgeNonValidWords(result);
```
هر کدوم از این خط‌ها پتانسیلش رو داره یه سرویس جدا بشه:)))
چرا؟

navidnasyry commented on FullTextSearch/Application/InvertedIndex/StringToWordsProcessor.cs:
```diff
@@ -0,0 +1,73 @@
+    private static bool IsLongEnough(string value) => value.Length >= 3;
+
+
+    private static bool IsNotNoise(string value) =>
+        !AppSettings.RegexPatterns.UrlRegex().IsMatch(value) &&
+        !AppSettings.RegexPatterns.EmailRegex().IsMatch(value) &&
+        !AppSettings.RegexPatterns.PhoneNumberRegex().IsMatch(value);
```
این تابع‌ها شدیدا منطق دارن
 پرایوت استاتیک شدنشون چه بدی ای داره؟

navidnasyry replied:
قابل تعمیم به بقیه تابع‌های پرایوت استاتیک

navidnasyry commented on FullTextSearch/Application/InvertedIndex/QuerySearcher.cs:
```diff
@@ -0,0 +1,73 @@
+internal sealed class QuerySearcher(IWordsProcessor wordsProcessor, ISearcher searcher)
+    : IQuerySearcher
+{
+    private readonly ISearcher                         _searcher       = searcher ?? throw new ArgumentNullException(nameof(searcher));
+    private readonly IWordsProcessor                   _wordsProcessor = wordsProcessor ?? throw new ArgumentNullException(nameof(wordsProcessor));
+    private          Dictionary<string, List<string>>? _invertedIndex;
+    private          bool                              _isConstructed;
+    private          string[]                          _queryWords = null!;
```
اینا اسپیساشون زیاده یا به من داره بد نشون میده؟:)

navidnasyry commented on FullTextSearch/Application/InvertedIndex/InvertedIndexDictionary.cs:
```diff
@@ -0,0 +1,35 @@
+{
+    private readonly IInvertedIndexDictionaryFiller _indexDictionaryFiller = invertedIndexDictionaryFiller
+                                                                          ?? throw new ArgumentNullException(nameof(invertedIndexDictionaryFiller));
+    private readonly IQuerySearcher                 _searcher              = querySearcher
+                                                                          ?? throw new ArgumentNullException(nameof(querySearcher));
+    private          bool                           _isConstructed;
+
```
جایی تو کلین کد دیدی گفته باشه اینقدر اسپیس بخوره؟

navidnasyry replied:
شایدم گیتهاب داره بد نشون میده

navidnasyry commented on FullTextSearch/AppSettings.cs:
```diff
@@ -0,0 +1,38 @@
+
+namespace FullTextSearch;
+
+internal static partial class AppSettings
```
چرا partial

Shafagh-Sepehr committed:
refactor: move services to the Services folder and move interfaces to a new Abstractions folder

Shafagh-Sepehr committed:
refactor: adjust all namespaces according to IDE suggestions.

Shafagh-Sepehr committed:
refactor(Words): rename and relocate DataViewModels folder and rename Words class

rename DataViewModels folder to Models and move it to Application folder.
rename Words class to QueryProcessedWords.

Shafagh-Sepehr committed:
refactor(Services): change behaviour for better usability.

Shafagh-Sepehr committed:
refactor(ConfigHolder): change ConfigHolder to interface and implementation and add it to Services.

Shafagh-Sepehr committed:
refactor(ConfigurationService): move ConfigurationService and its abstraction to a new ConfigurationService folder.

Shafagh-Sepehr committed:
refactor(Services): rename Services to ServiceCollection to eliminate ambiguity

Shafagh-Sepehr committed:
refactor(StringToWordsProcessor): extract regex methods to new services

Shafagh-Sepehr committed:
refactor(StringToWordsProcessor): extract PurgeNoise function to its own service.

Shafagh-Sepehr committed:
refactor(StringToWordsProcessor): extract CleanAndSplit method to its own service.

Shafagh-Sepehr committed:
refactor(StringToWordsProcessor): extract CleanAndSelect method to its own service.

Shafagh-Sepehr committed:
refactor(StringToWordsProcessor): extract Stem method to its own service.

Shafagh-Sepehr committed:
refactor(StringCleaners): rename services to better represent their action.

Shafagh-Sepehr committed:
refactor(StringToWordsProcessor): extract PurgeNonValidWords method to its own service.
